File: spark/sparkdemo/sparkcore/src/main/java/com/itguigu/wc/WordCount.scala
package com.itguigu.wc

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

/**
 * Project: sparkdemo
 * Package: com.itguigu.wc
 * Version: 1.0
 * Created by ljy on 2021-11-27 21:34
 */

/**
 * @ClassName ${CLASSNAME}
 * @Author: ljy on 2021-11-27 21:34
 * @Version: 1.0
 * @Description:字符统计
 */
object WordCount {
  def main(args: Array[String]): Unit = {
    //TODO: 创建Spark运行配置对象
    val conf = new SparkConf().setAppName("WordCount")
      .setMaster("local[*]")
      .set("spark.executor.memory","2g")


    //TODO: 创建Spark上下文环境对象（连接对象）
    val sc:SparkContext = new SparkContext(conf)

    //TODO: 读取文件数据
    val fileRDD: RDD[String] = sc.textFile("input/word.txt")

    //TODO： 将文件中的数据进行分词
    val wordRDD:RDD[String] = fileRDD.flatMap(_.split(","))

    //TODO: 转换数据结构word => （word,1）
    val word2OneRDD:RDD[(String,Int)] = wordRDD.map((_, 1))

    //TODO: 将转换结构后的数据安装相同的单词进行分组聚合
    val word2CountRDD:RDD[(String,Int)] = word2OneRDD.reduceByKey(_ + _)

    //TODO: 将数据聚合结果采集到内存中
    val word2Count:Array[(String,Int)] = word2CountRDD.collect()

    //TODO: 打印结果
    word2Count.foreach(println)

    //TODO: 关闭Spark连接
    sc.stop()
  }
}

File: spark/spark_java/src/main/java/com/ljy/wc/WordCount.java
package com.ljy.wc;/**
 * Project: spark_java
 * Package: com.ljy.test
 * Version: 1.0
 * Created by ljy on 2021-12-23 11:00
 */

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.MapFunction;
import scala.Int;
import scala.Tuple2;


import java.util.Arrays;
import java.util.Iterator;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

/**
 * @ClassName Main
 * @Author: ljy on 2021-12-23 11:00
 * @Version: 1.0
 * @Description:测试java的spark开发环境
 */
public class WordCount {
    public static void main(String[] args) {

        SparkConf conf = new SparkConf();
        conf.setMaster("local[*]");
        conf.setAppName("spark_wordcount_java");

        JavaSparkContext sc = new JavaSparkContext(conf);
        JavaRDD<String> lines = sc.textFile("./input/word.txt");
        JavaRDD<String> words = lines.flatMap(line -> Arrays.asList(line.split(",")).iterator());

        //使用group ->map
        /*JavaPairRDD<String, Iterable<String>> wordGroup = words.groupBy(new Function<String, String>() {

            @Override
            public String call(String s) throws Exception {
                return s;
            }
        });

        JavaRDD<Tuple2<String, Integer>> result = wordGroup.map(new Function<Tuple2<String, Iterable<String>>, Tuple2<String, Integer>>() {

            @Override
            public Tuple2<String, Integer> call(Tuple2<String, Iterable<String>> stringIterableTuple2) throws Exception {
                return new Tuple2<String, Integer>(stringIterableTuple2._1, StreamSupport.stream(stringIterableTuple2._2.spliterator(), false).collect(Collectors.toList()).size());
            }
        });*/

        //1. 使用reducekey
        JavaPairRDD<String, Integer> wordToCount = words.mapToPair(word -> new Tuple2<>(word, 1));

        //使用reduceByKey
        JavaPairRDD<String, Integer> result = wordToCount.reduceByKey((v1, v2) -> (v1 + v2));//


        result.foreach(tuple2 -> System.out.println(tuple2));

    }
}
